util/buffer.go

```diff
@@ -0,0 +1,140 @@
+package util
```

Comment:
I don't think a "util" package is a good idea, and would prefer something like `internal/buffer` instead.  Then the constructor becomes `buffer.New()`.

As mentioned previously, I also think that the buffer type should only be an implementation detail within the package, and that any exported marshaling and unmarshaling functions should use the stdlib interfaces, even if they use a buffer underneath.